### ISSUE TITLE
wiki does not have any advice/suggestions for what to do if applesmc cannot be loaded

### ISSUE THREAD

**Issue by katbyte** [t2linux/wiki#353]
Unable to load applesmc on a Macmini8,1 (A1993) for fan control

```
[10:09:28] root@mini:~/src/mbpfan▸master# modprobe applesmc
modprobe: ERROR: could not insert 'applesmc': Input/output error
[10:11:26] root@mini:~/src/mbpfan▸master# 
```

dmesg:
```
[   61.148266] applesmc: probe of applesmc.768 failed with error -5
[  118.852271] applesmc: driver init failed (ret=-5)!
```

Linux version: `Linux version 5.19.7-2-pve (build@proxmox) (gcc (Debian 10.2.1-6) 10.2.1 20210110, GNU ld (GNU Binutils for Debian) 2.35.2) #1 SMP PREEMPT_DYNAMIC PVE 5.19.7-2 (Tue, 04 Oct 2022 17:18:40 +`

Any ideas of what could be causing this/how I can fix it? I have reset SMC (hold cmd+opt+r+p till 2nd chime)

**Comment by AdityaGarg8** [comment]
You need a patched kernel having the T2 patches.

https://github.com/andersfugmann/T2-Debian-Kernel

It also has patches for Wifi.

**Comment by katbyte** [comment]
hmm interesting, so no way to control fans on a T2 mac without patching the kernel then?

I wonder if this will work on the proxmox kernel 🤔

**Comment by AdityaGarg8** [comment]
Well patches are essentially the drivers which are yet to be upstreamed. Obviously you cannot use your hardware without drivers for the same.

**Comment by katbyte** [comment]
apologies for my lack of understanding linux drivers: is applesmc.ko considered part of the kernel and those patches are applied to the kernel source to update it? Or at they applied to the kernel which we then load to support the applesmc.ko?

hmm seems like i could download the proxmox kernel source and potentially modify/run `t2-kernel-script-debian` to patch and build it

**Comment by AdityaGarg8** [comment]
Apple SMC had been patches to support the T2 Macs in the kernel. In the upstream version, the driver has supported for only the older Macs.

https://github.com/AdityaGarg8/linux-t2-patches

This is the complete patchset used to patch the kernels.

**Comment by katbyte** [comment]
oh perfect thanks! 

been a very long time since i've looked at a device driver but the fan control one seems straight forward so i'll take a stab at applying it to the proxmox kernel

**Comment by AdityaGarg8** [comment]
You probably could edit the build.sh in https://github.com/t2linux/T2-Ubuntu-Kernel to compile your kernel. Although I dunno how a proxmox kernel works/compiles

**Comment by AdityaGarg8** [comment]
Due t demands of some more users, an initial test repo for proxmox has been made. Would you like to test kernels for proxmox @katbyte ?

**Comment by jacob9423** [comment]
@AdityaGarg8 Hello. I just wanted to let you know that I was also looking for this fix as well.

I went ahead and installed the kernel you were referencing too and now the applesmc module is loading and I was able to get mbpfan installed (I had to install make and gcc to get it to install though because the default proxmox install doesn't have those it seems)

The kernal I installed was the 5.19 version.
